app/models/workers_scaling/algorithm_runner.rb

```diff
@@ -3,96 +3,69 @@ module WorkersScaling
   ##
   # Scheduler to run Algorithm decision loop periodically
   class AlgorithmRunner
-    @@cache = {}
 
-    ##
-    # Returns instance of AlgorithmRunner for experiment_id if registered
-    # Otherwise returns nil
-    def self.get(experiment_id)
-      LOGGER.debug "Getting runner from cache, experiment id: #{experiment_id}"
-      return @@cache[experiment_id.to_s] if @@cache
-    end
+    RUNNER_INTERVAL = 15.seconds
+    THREADS_NUMBER = 4
 
     ##
-    # Registers instance of AlgorithmRunner for experiment_id
-    def self.put(experiment_id, runner)
-      @@cache[experiment_id.to_s] = runner
-      LOGGER.debug "Adding runner to cache, experiment id: #{experiment_id}"
-    end
-
-    ##
-    # Unregisters instance of AlgorithmRunner for experiment_id
-    def self.delete(experiment_id)
-      @@cache.delete(experiment_id.to_s) if @@cache
-      LOGGER.debug "Deleting runner from cache, experiment id: #{experiment_id}"
-    end
-
-    ##
-    # Expects:
-    # * instance of Experiment
-    # * instance of Algorithm for given Experiment
-    # * interval between decision loop executions in seconds
-    def initialize(experiment, algorithm, interval)
-      @experiment = experiment
-      @algorithm = algorithm
-      @interval = interval
-      @mutex = Mutex.new
-
-      self.class.put(@experiment.id, self)
-    end
-
-    ##
-    # Initializes Algorithm, then starts its decision loop
-    # Stops when there is no next execution time set
-    def start
+    # Runs periodically and loads some data from all saved Algorithms
+    # If next_execution_time has passed, starts execute_and_schedule for this algorithm
+    def self.start
       Thread.new do
         begin
-          @algorithm.initial_deployment
-          @next_execution_time = Time.now + @interval
-
-          @mutex.synchronize do
-            catch :finished do
-              until @next_execution_time.nil?
+          while true do
+            LOGGER.debug 'Entering Algorithm Runner loop'
+            work_queue = Queue.new
+            Algorithm.get_all_algorithms_times.each do |experiment_id, next_execution_time|
+              if next_execution_time <= Time.now
+                work_queue.push(experiment_id)
+              end
+            end
 
-                while @next_execution_time > Time.now do
-                  @mutex.sleep @next_execution_time - Time.now
-                  throw :finished if @next_execution_time.nil?
+            threads = (1..THREADS_NUMBER).map do
+              Thread.new do
+                until work_queue.empty?
+                  execute_and_schedule(work_queue.pop)
```

Comment:
Rzeczywiście, może tak być. W takim razie są dwie opcje:
# Dodać mutex chroniący kolejkę - trzeba będzie trochę przerobić całą pętlę.
# Zamiast pytać czy .empty? od razu robić pop(non_block=false), które spowoduje wyrzucenie wyjątku, jeżeli kolejka jest pusta.
# Najciekawsza według mnie opcja - jeżeli zawołamy pop na pustej kolejce, wołający wątek usypia. Można na samym początku odpalić te 4 wątki i nie zabijać ich - jeżeli skończą im się zadania to zwyczajnie usną. Zniknie przy okazji narzut na uruchamianie za każdym razem nowych.

Czekam na opinie.